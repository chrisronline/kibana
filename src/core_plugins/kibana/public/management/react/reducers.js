/* eslint-disable */
import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux'

import app from './app/app.reducer';
import indexPattern from './index-pattern/index-pattern.reducer';

export const rootReducer = combineReducers({
  routing: routerReducer,
  app,
  indexPattern,
});
