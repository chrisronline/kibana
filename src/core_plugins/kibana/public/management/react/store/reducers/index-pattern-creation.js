import { handleActions } from 'redux-actions';
import { createSelector } from 'reselect';
import { chunk, sortBy as sortByLodash } from 'lodash';

import {
  selectTimeField,
  fetchedTimeFields,
  goToNextPage,
  goToPreviousPage,
  changeSort,
  fetchedIndices,
  includeSystemIndices,
  excludeSystemIndices,
  creatingIndexPattern,
  createdIndexPattern,
} from '../actions/index-pattern-creation';

import {
  getIndexPatternCreate,
} from '../../reducers';

const defaultState = {
  isIncludingSystemIndices: false,
  isCreating: false,
  timeFields: {
    timeFields: undefined,
    selectedTimeField: undefined,
  },
  results: {
    indices: undefined,
    pattern: undefined,
    hasExactMatches: false,
    perPage: 10,
    page: 1,
    sortBy: undefined,
    sortAsc: true,
  }
}

export default handleActions({
  [selectTimeField](state, { payload }) {
    const { timeFields } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        timeFields,
      },
    };
  },
  [fetchedTimeFields](state, { payload }) {
    const { timeFields } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        timeFields,
      },
    };
  },
  [selectTimeField](state, { payload }) {
    const { timeField } = payload;

    return {
      ...state,
      timeFields: {
        ...state.timeFields,
        selectedTimeField: timeField,
      },
    };
  },
  [fetchedIndices](state, { payload }) {
    const { indices, hasExactMatches, pattern } = payload;

    return {
      ...state,
      results: {
        ...state.results,
        indices,
        hasExactMatches,
        pattern,
      },
    };
  },
  [changeSort](state, { payload }) {
    const { sortBy, sortAsc } = payload;

    return {
      ...state,
      results: {
        ...state.results,
        sortBy,
        sortAsc: sortBy === state.results.sortBy ? !state.results.sortAsc: sortAsc,
      },
    };
  },
  [goToNextPage](state, { payload }) {
    return {
      ...state,
      results: {
        ...state.results,
        page: getNextPage(state.results),
      },
    };
  },
  [goToPreviousPage](state, { payload }) {
    return {
      ...state,
      results: {
        ...state.results,
        page: getPreviousPage(state.results),
      },
    };
  },
  [includeSystemIndices](state, action) {
    return {
      ...state,
      isIncludingSystemIndices: true,
    };
  },
  [excludeSystemIndices](state, action) {
    return {
      ...state,
      isIncludingSystemIndices: false,
    };
  },
  [creatingIndexPattern](state, action) {
    return {
      ...state,
      isCreating: true,
    };
  },
  [createdIndexPattern](state, action) {
    return {
      ...state,
      isCreating: false,
    };
  },
}, defaultState);

function getNextPage({ page, perPage, indices }) {
  const pages = Math.ceil(indices.length / perPage);
  const nextPage = page + 1;
  return nextPage > pages
    ? 1
    : nextPage;
}

function getPreviousPage({ page, perPage, indices }) {
  const pages = Math.ceil(indices.length / perPage);
  const previousPage = page - 1;
  return previousPage < 1
    ? pages
    : previousPage;
}

function getFilteredIndices(indices, isIncludingSystemIndices) {
  if (!indices) {
    return indices;
  }

  if (isIncludingSystemIndices) {
    return indices;
  }

  // All system indices begin with a period.
  return indices.filter(index => !index.name.startsWith('.'));
}

function getPaginatedIndices(indices, page, perPage) {
  const pagesOfIndices = chunk(indices, perPage);
  return pagesOfIndices[page - 1];
}

export const getFilteredAndPaginatedIndices = createSelector(
  [
    state => state.results,
    state => state.isIncludingSystemIndices
  ],
  ({
    indices: allIndices,
    page,
    perPage,
    sortBy,
    sortAsc,
  }, isIncludingSystemIndices) => {
    if (allIndices === undefined) {
      return {
        indices: undefined,
        numOfPages: 0,
      };
    }

    let filteredIndices = getFilteredIndices(allIndices, isIncludingSystemIndices);
    if (!!sortBy) {
      filteredIndices = sortByLodash(filteredIndices, sortBy);
      if (!sortAsc) {
        filteredIndices.reverse();
      }
    }

    const numOfPages = Math.ceil(filteredIndices.length / perPage);
    const indices = getPaginatedIndices(filteredIndices, page, perPage);
    return {
      indices,
      numOfPages,
    }
  }
);

export const getPattern = state => getIndexPatternCreate(state).results.pattern;
export const getSelectedTimeField = state => getIndexPatternCreate(state).timeFields.selectedTimeField;
