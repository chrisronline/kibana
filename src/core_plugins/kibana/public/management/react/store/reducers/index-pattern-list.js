import { handleActions } from 'redux-actions';

import {
  fetchedIndexPatterns,
  setTransientTableId,
} from '../actions/index-pattern-list';

import {
  getIndexPatternList,
} from '../../reducers';

const defaultState = {
  indexPatterns: undefined,
  transientTableId: undefined,
};

export default handleActions({
  [fetchedIndexPatterns](state, { payload }) {
    const { indexPatterns } = payload;

    return {
      ...state,
      indexPatterns,
    };
  },
  [setTransientTableId](state, { payload: { id: transientTableId } }) {
    return {
      ...state,
      transientTableId,
    };
  }
}, defaultState);

export const getIndexPatterns = state => getIndexPatternList(state).indexPatterns;
export const getPathToIndexPatterns = () => 'indexPatterns';
