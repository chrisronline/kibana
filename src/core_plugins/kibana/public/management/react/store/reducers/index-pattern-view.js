import { handleActions } from 'redux-actions';

import {
  fetchedIndexPattern,
  setAsDefaultIndexPattern,
} from '../actions/index-pattern-view';

import {
  getIndexPatternView,
} from '../../reducers';

const defaultState = {
  indexPattern: {
    pattern: undefined,
    fields: undefined,
    timeFieldName: undefined,
    isDefault: undefined,
    id: undefined,
  },
  fieldsTable: {},
  tabs: {},
};

export const indexPatternView = handleActions({
  [fetchedIndexPattern](state, { payload: { indexPattern } }) {
    return {
      ...state,
      indexPattern,
    };
  },
  [setAsDefaultIndexPattern](state) {
    return {
      ...state,
      indexPattern: {
        ...state.indexPattern,
        isDefault: true,
      }
    };
  }
}, defaultState);

export const getPathToFields = () => 'indexPattern.fields';
export const getTabs = state => getIndexPatternView(state).tabs;
