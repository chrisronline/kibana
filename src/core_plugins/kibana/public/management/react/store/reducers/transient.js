import { handleActions } from 'redux-actions';

import {
  change,
  createItem,
} from '../actions/transient';

const defaultState = {};

export default handleActions({
  [createItem](state, { payload: { id, defaultData }}) {
    return {
      ...state,
      [id]: defaultData || {},
    }
  },
  [change](state, { payload: { id, data } }) {
    return {
      ...state,
      [id]: {
        ...state[id],
        ...data
      },
    };
  },
}, defaultState);
