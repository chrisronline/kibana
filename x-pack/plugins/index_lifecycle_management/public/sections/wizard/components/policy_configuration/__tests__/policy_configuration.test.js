/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import { PolicyConfiguration } from '../';
import { indexLifecycleManagement } from '../../../../../store/reducers';
import { PHASE_HOT, PHASE_WARM, PHASE_COLD, PHASE_DELETE, PHASE_ROLLOVER_ALIAS } from '../../../../../store/constants';

jest.mock('@elastic/eui', () => ({
  EuiTitle: props => props.children || '',
  EuiText: props => props.children || '',
  EuiSpacer: props => props.children || '',
  EuiHorizontalRule: props => props.children || '',
  EuiButton: props => props.children || '',
  EuiButtonEmpty: props => props.children || '',
}));

jest.mock('../../../../../store/actions', () => ({
  setBootstrapEnabled: jest.fn(),
  setIndexName: jest.fn(),
  setAliasName: jest.fn(),
  setSelectedPolicyName: jest.fn(),
  setSaveAsNewPolicy: jest.fn(),
}));

jest.mock('../../policy_selection', () => ({
  PolicySelection: props => props.children || ''
}));
jest.mock('../../node_attrs_details', () => ({
  NodeAttrsDetails: props => props.children || ''
}));

jest.mock('../components/hot_phase', () => ({
  HotPhase: props => props.children || ''
}));
jest.mock('../components/warm_phase', () => ({
  WarmPhase: props => props.children || ''
}));
jest.mock('../components/cold_phase', () => ({
  ColdPhase: props => props.children || ''
}));
jest.mock('../components/delete_phase', () => ({
  DeletePhase: props => props.children || ''
}));

const defaultErrors = {
  [PHASE_HOT]: {},
  [PHASE_WARM]: {},
  [PHASE_COLD]: {},
  [PHASE_DELETE]: {},
};

const defaultStore = {
  subscribe: jest.fn(),
  dispatch: jest.fn(),
};

describe('PolicyConfiguration', () => {
  it('should render just the policy selection if no policy is selected yet', () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <PolicyConfiguration
          validate={() => {}}
          errors={defaultErrors}
          done={() => {}}
          back={() => {}}
        />
      </Provider>
    );

    expect(component.find('PolicyConfiguration').children()).toMatchSnapshot();
  });

  it('should render the selected new policy', () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        },
        policies: {
          ...defaultState.policies,
          selectedPolicySet: true,
          selectedPolicy: {
            saveAsNew: true,
            name: 'my_policy'
          }
        }
      }),
      ...defaultStore,
    };

    const component = mount(
      <Provider store={store}>
        <PolicyConfiguration
          validate={() => {}}
          errors={defaultErrors}
          done={() => {}}
          back={() => {}}
        />
      </Provider>
    );

    expect(component.find('PolicyConfiguration')).toMatchSnapshot();
  });

  it('should return error state on validation', async () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        },
        policies: {
          ...defaultState.policies,
          selectedPolicySet: true,
          selectedPolicy: {
            saveAsNew: true,
            name: 'my_policy'
          }
        }
      }),
      ...defaultStore,
    };
    const validate = jest.fn();
    const component = mount(
      <Provider store={store}>
        <PolicyConfiguration
          validate={validate}
          errors={{
            ...defaultErrors,
            [PHASE_HOT]: {
              [PHASE_ROLLOVER_ALIAS]: ['There is an error.'],
            }
          }}
          done={() => {}}
          back={() => {}}
        />
      </Provider>
    );

    const result = await component.find('PolicyConfiguration').instance().validate();
    expect(result).toBe(false);
    expect(validate).toHaveBeenCalled();
  });

  it('should submit if valid', async () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        },
        policies: {
          ...defaultState.policies,
          selectedPolicySet: true,
          selectedPolicy: {
            saveAsNew: true,
            name: 'my_policy'
          }
        }
      }),
      ...defaultStore,
    };
    const done = jest.fn();
    const component = mount(
      <Provider store={store}>
        <PolicyConfiguration
          validate={() => {}}
          errors={{
            ...defaultErrors,
            [PHASE_WARM]: {
              [PHASE_ROLLOVER_ALIAS]: ['There is an error.'],
            }
          }}
          done={done}
          back={() => {}}
        />
      </Provider>
    );

    await component.find('PolicyConfiguration').instance().submit();
    expect(done).not.toHaveBeenCalled();

    const noErrorComponent = mount(
      <Provider store={store}>
        <PolicyConfiguration
          validate={() => {}}
          errors={{
            ...defaultErrors,
            [PHASE_WARM]: {
              [PHASE_ROLLOVER_ALIAS]: [],
            }
          }}
          done={done}
          back={() => {}}
        />
      </Provider>
    );

    await noErrorComponent.find('PolicyConfiguration').instance().submit();
    expect(done).toHaveBeenCalled();
  });

  it('should render the flyout', () => {
    const defaultState = indexLifecycleManagement({}, {});
    const store = {
      getState: () => ({
        ...defaultState,
        indexTemplate: {
          ...defaultState.indexTemplate,
          indexTemplates: [{}]
        },
        policies: {
          ...defaultState.policies,
          selectedPolicySet: true,
          selectedPolicy: {
            saveAsNew: true,
            name: 'my_policy'
          }
        }
      }),
      ...defaultStore,
    };
    const component = mount(
      <Provider store={store}>
        <PolicyConfiguration
          validate={() => {}}
          errors={defaultErrors}
          done={() => {}}
          back={() => {}}
        />
      </Provider>
    );

    component.find('PolicyConfiguration').instance().showNodeDetailsFlyout('foo');
    component.update();

    expect(component.find('PolicyConfiguration').find('NodeAttrsDetails')).toMatchSnapshot();
  });
});
