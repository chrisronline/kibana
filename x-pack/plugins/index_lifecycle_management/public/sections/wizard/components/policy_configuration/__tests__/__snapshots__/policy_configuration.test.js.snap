// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicyConfiguration should render just the policy selection if no policy is selected yet 1`] = `<PolicySelection />`;

exports[`PolicyConfiguration should render the flyout 1`] = `
<NodeAttrsDetails
  close={[Function]}
  selectedNodeAttrs="foo"
/>
`;

exports[`PolicyConfiguration should render the selected new policy 1`] = `
<PolicyConfiguration
  back={[Function]}
  dispatch={[MockFunction]}
  done={[Function]}
  errors={
    Object {
      "cold": Object {},
      "delete": Object {},
      "hot": Object {},
      "warm": Object {},
    }
  }
  isSelectedPolicySet={true}
  selectedPolicyName="my_policy"
  validate={[Function]}
>
  <div
    className="euiAnimateContentLoad"
  >
    <PolicySelection />
    <EuiHorizontalRule
      className="ilmHrule"
    />
    <EuiTitle>
      <h4>
        Edit policy my_policy
      </h4>
    </EuiTitle>
    <EuiSpacer
      size="xs"
    />
    <EuiText
      color="subdued"
    >
      <p>
        Configure the phases of your data and when to transition between them.
      </p>
    </EuiText>
    <EuiSpacer />
    <HotPhase
      errors={Object {}}
      isShowingErrors={false}
      validate={[Function]}
    />
    <EuiHorizontalRule
      className="ilmHrule"
    />
    <WarmPhase
      errors={Object {}}
      isShowingErrors={false}
      showNodeDetailsFlyout={[Function]}
      validate={[Function]}
    />
    <EuiHorizontalRule
      className="ilmHrule"
    />
    <ColdPhase
      errors={Object {}}
      isShowingErrors={false}
      showNodeDetailsFlyout={[Function]}
      validate={[Function]}
    />
    <EuiHorizontalRule
      className="ilmHrule"
    />
    <DeletePhase
      errors={Object {}}
      isShowingErrors={false}
      validate={[Function]}
    />
    <EuiHorizontalRule
      className="ilmHrule"
    />
    <EuiButtonEmpty
      iconSide="left"
      iconType="sortLeft"
      onClick={[Function]}
    >
      Back
    </EuiButtonEmpty>
      
    <EuiButton
      fill={true}
      iconSide="right"
      iconType="sortRight"
      onClick={[Function]}
    >
      Continue
    </EuiButton>
  </div>
</PolicyConfiguration>
`;
